nes-network/bindings/network/lib.rs

```diff
@@ -0,0 +1,359 @@
+    let buffer = match receiver_channel.chan.receive() {
+        ReceiverChannelResult::Ok(buffer) => buffer,
+        ReceiverChannelResult::Closed => {
+            return Ok(false);
+        }
+        ReceiverChannelResult::Error(e) => {
+            return Err(format!("Error while receiving buffer: {}", e).into());
+        }
+    };
+
+    builder.as_mut().setMetadata(&ffi::SerializedTupleBuffer {
+        sequence_number: buffer.sequence_number as usize,
+        origin_id: buffer.origin_id as usize,
+        watermark: buffer.watermark as usize,
+        chunk_number: buffer.chunk_number as usize,
+        number_of_tuples: buffer.number_of_tuples as usize,
+        last_chunk: buffer.last_chunk,
+    });
+
+    builder.as_mut().setData(&buffer.data);
+
+    for child_buffer in buffer.child_buffers.iter() {
+        assert!(!child_buffer.is_empty());
+        builder.as_mut().addChildBuffer(child_buffer);
+    }
+
+    Ok(true)
+}
+
+// The `interrupt_receive` and `close_receiver_channel` are identical because,
+// currently, there is no requirement to keep the channel alive after a `receive_buffer` has been
+// interrupted as the network source will shut down and never try to receive a buffer again.
+// And closing the channel is the easiest option to interrupt a pending `receive_buffer`.
+fn interrupt_receive(channel: &ReceiverDataChannel) {
+    channel.chan.close();
+}
+
+// CXX requires the usage of Boxed types
+#[allow(clippy::boxed_local)]
+fn close_receiver_channel(channel: Box<ReceiverDataChannel>) {
+    channel.chan.close();
+}
+
+fn register_sender_channel(
+    sender_service: &SenderNetworkService,
+    connection_addr: String,
+    channel_id: String,
+) -> Result<Box<SenderDataChannel>, String> {
+    let connection_addr =
+        ConnectionIdentifier::from_str(connection_addr.as_str()).map_err(|e| e.to_string())?;
+    let data_queue = match &sender_service.handle {
+        SenderService::MemCom(s) => s.register_channel(connection_addr.clone(), channel_id.clone()),
+        SenderService::Tcp(s) => s.register_channel(connection_addr.clone(), channel_id.clone()),
+    }
+        .unwrap();
+
+    Ok(Box::new(SenderDataChannel { chan: data_queue }))
+}
+
+fn try_send_on_channel(
+    channel: &SenderDataChannel,
+    metadata: ffi::SerializedTupleBuffer,
+    data: &[u8],
+    children: &[&[u8]],
+) -> ffi::SendResult {
+    let buffer = TupleBuffer {
+        sequence_number: metadata.sequence_number as u64,
+        origin_id: metadata.origin_id as u64,
+        chunk_number: metadata.chunk_number as u64,
+        number_of_tuples: data.len() as u64,
+        watermark: metadata.watermark as u64,
+        last_chunk: metadata.last_chunk,
+        data: Vec::from(data),
+        child_buffers: children.iter().map(|bytes| Vec::from(*bytes)).collect(),
+    };
+
+    // Because we copy the data anyway, we don't have to reuse the buffer if sending failed.
+    match channel.chan.try_send_data(buffer) {
```

Comment:
The layers of indirection are a bit confusing (channel contains chan), which I don't think is necessary here.